mava/systems/ppo/sebulba/ff_mappo.py

```diff
@@ -0,0 +1,709 @@
+                for source in params_sources:
+                    source.update(params)
+
+        # Pass all the metrics and  params to the main thread (evaluator) for logging and evaluation
+        ep_metrics, train_metrics = tree.map(lambda *x: np.asarray(x), *metrics)
+        rollout_times: Dict[str, NDArray] = tree.map(lambda *x: np.mean(x), *rollout_times_array)
+        timing_dict = rollout_times | learn_times
+        timing_dict = tree.map(np.mean, timing_dict, is_leaf=lambda x: isinstance(x, list))
+
+        eval_queue.put((ep_metrics, train_metrics, learner_state, timing_dict))
+
+
+def learner_setup(
+    key: chex.PRNGKey, config: DictConfig, learner_devices: List
+) -> Tuple[
+    SebulbaLearnerFn[SebulbaLearnerState, PPOTransition],
+    Tuple[ActorApply, CriticApply],
+    SebulbaLearnerState,
+    Sharding,
+]:
+    """Initialise learner_fn, network and learner state."""
+
+    # Create temporory envoirnments.
+    env = environments.make_gym_env(config, config.arch.num_envs, add_global_state=True)
+    # Get number of agents and actions.
+    action_space = env.single_action_space
+    config.system.num_agents = len(action_space)
+    config.system.num_actions = int(action_space[0].n)
+
+    devices = mesh_utils.create_device_mesh((len(learner_devices),), devices=learner_devices)
+    mesh = Mesh(devices, axis_names=("learner_devices",))
+    model_spec = PartitionSpec()
+    data_spec = PartitionSpec("learner_devices")
+    learner_sharding = NamedSharding(mesh, model_spec)
+
+    # PRNG keys.
+    key, actor_key, critic_key = jax.random.split(key, 3)
+
+    # Define network and optimiser.
+    actor_torso = hydra.utils.instantiate(config.network.actor_network.pre_torso)
+    action_head, _ = get_action_head(action_space)
+    actor_action_head = hydra.utils.instantiate(action_head, action_dim=config.system.num_actions)
+
+    critic_torso = hydra.utils.instantiate(config.network.critic_network.pre_torso)
+
+    actor_network = Actor(torso=actor_torso, action_head=actor_action_head)
+    critic_network = Critic(torso=critic_torso, centralised_critic=True)
+
+    actor_lr = make_learning_rate(config.system.actor_lr, config)
+    critic_lr = make_learning_rate(config.system.critic_lr, config)
+
+    actor_optim = optax.chain(
+        optax.clip_by_global_norm(config.system.max_grad_norm),
+        optax.adam(actor_lr, eps=1e-5),
+    )
+    critic_optim = optax.chain(
+        optax.clip_by_global_norm(config.system.max_grad_norm),
+        optax.adam(critic_lr, eps=1e-5),
+    )
+
+    # Initialise observation: Select only obs for a single agent.
+    init_obs = env.reset().observation
+    init_agents_view = init_obs.agents_view[0][jnp.newaxis, :]
+    init_action_mask = init_obs.action_mask[0][jnp.newaxis, :]
+    init_global_state = init_obs.global_state[0][jnp.newaxis, :]  # type: ignore
+    init_x = ObservationGlobalState(init_agents_view, init_action_mask, init_global_state)
```

Comment:
I missed this on the first go round, but spotted it in the other PR, we gotta find a way to return a global state spec